### PR TITLE
RFS-327 - Dynamic gas cost for getBtcTransactionConfirmations

### PR DESCRIPTION
**Pull request by amendelzon**
This PR sets the (dynamic) gas cost for the new Bridge method getBtcTransactionConfirmations. It also adds a few extra estimation tests that helped in the actual dynamic function calculation.

The cost calculation and its error handling are also unit tested.

The cost itself is based upon an interpolation of two cases (zero confirmations vs. one hundred confirmations), each averaged from 2000 random cases. Details on these tests can be seen [here](https://docs.google.com/spreadsheets/d/1c4b2vCcpCKVhZ7opEH_OlLn8LIeiGPyRdBtjL1xJSKA/edit?usp=sharing).

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 41 issues
* ![MAJOR][MAJOR] 17 major
* ![MINOR][MINOR] 15 minor
* ![INFO][INFO] 9 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [Bridge.java#L582](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L582): Method co.rsk.peg.Bridge.getBtcTransactionConfirmationsGetCost(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L592](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L592): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L594](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L594): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L596](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L596): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L678](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L678): Method co.rsk.peg.Bridge.getFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L751](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L751): Method co.rsk.peg.Bridge.getRetiringFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L814](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L814): Method co.rsk.peg.Bridge.addFederatorPublicKeyMultikey(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L815](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L815): Method co.rsk.peg.Bridge.addFederatorPublicKeyMultikey(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L895](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L895): Method co.rsk.peg.Bridge.getPendingFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L346](https://github.com/rsksmart/rskj/blob/96072be6148edc1b205f59b8e93f047093e74239/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L346): The Cyclomatic Complexity of this method "registerBtcTransaction" is 17 which is greater than 10 authorized. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by josedahlquist** [comment]
Besides my thumb up, I have some concerns about the change for the bridge setup on the init... Could it affect performance somehow? (I recall the bridgeSupport initialization was kinda heavy in cpu time)
Can't think of a better place to do it now that you need it for the gas calculation though... just raising the hand to see if someone else see this as a potential issue and can think on a better solution.

**Comment by amendelzon** [comment]
> Besides my thumb up, I have some concerns about the change for the bridge setup on the init... Could it affect performance somehow? (I recall the bridgeSupport initialization was kinda heavy in cpu time)
> Can't think of a better place to do it now that you need it for the gas calculation though... just raising the hand to see if someone else see this as a potential issue and can think on a better solution.

In some recent profiling I've been doing for the bridge in the context of some optimizations, the bridge setup itself only takes less than 2% of the time in the optimized version. Plus, unless we're only estimating gas, the bridge setup would execute sooner or later. So I think there should be no major concerns on this.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 1]
```diff
@@ -1132,6 +1132,40 @@ public Sha256Hash getBtcBlockchainBlockHashAtDepth(int depth) throws BlockStoreE
...
+            // Block not found, default to basic cost
+            if (block == null) {
+                return BASIC_COST;
+            }
+
+            final int bestChainHeight = getBtcBlockchainBestChainHeight();
+            final int blockDepth = bestChainHeight - block.getHeight();
+
+            // Block too deep, default to basic cost
+            if (blockDepth > BTC_TRANSACTION_CONFIRMATION_MAX_DEPTH) {
```
Comment (juli) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: Retrieving old blocks from disk (cache is limited to 65k items) is cheaper than traversing BTC_TRANSACTION_CONFIRMATION_MAX_DEPTH blocks in RAM. Is this correct?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 2]
```diff
@@ -1132,6 +1132,40 @@ public Sha256Hash getBtcBlockchainBlockHashAtDepth(int depth) throws BlockStoreE
...
+            this.ensureBtcBlockStore();
+            final StoredBlock block = btcBlockStore.getFromCache(btcBlockHash);
+
+            // Block not found, default to basic cost
+            if (block == null) {
+                return BASIC_COST;
+            }
+
+            final int bestChainHeight = getBtcBlockchainBestChainHeight();
+            final int blockDepth = bestChainHeight - block.getHeight();
```
Comment (juli) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: I recommend checking result is not < 0 here.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 3]
```diff
@@ -1132,6 +1132,40 @@ public Sha256Hash getBtcBlockchainBlockHashAtDepth(int depth) throws BlockStoreE
         return current.getHeader().getHash();
     }
 
+    public Long getBtcTransactionConfirmationsGetCost(Sha256Hash btcBlockHash) {
```
Comment (juli) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: Current implementation of `getBtcTransactionConfirmations` has a execution time proportional to the length of the array of transaction hashes, if a limit is not set then this method should consider that parameter.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 1 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: I'm not entirely sure on what do you mean or where does this question lead. Regardless:

- Retrieving one single block from disk is probably cheaper than retrieving BTC_TRANSACTION_CONFIRMATION_MAX_DEPTH (4320 at the time of writing) blocks from RAM, but I would have to check and get back to you. 
- Retrieving a given number of blocks from RAM should be cheaper than retrieving that same number of blocks from disk.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 2 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java]: :+1:

**Comment by rskops** [comment]
SonarQube analysis reported 41 issues
* ![MAJOR][MAJOR] 16 major
* ![MINOR][MINOR] 16 minor
* ![INFO][INFO] 9 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [Bridge.java#L582](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L582): Method co.rsk.peg.Bridge.getBtcTransactionConfirmationsGetCost(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L592](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L592): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L594](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L594): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L596](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L596): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L678](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L678): Method co.rsk.peg.Bridge.getFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L751](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L751): Method co.rsk.peg.Bridge.getRetiringFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L814](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L814): Method co.rsk.peg.Bridge.addFederatorPublicKeyMultikey(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L815](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L815): Method co.rsk.peg.Bridge.addFederatorPublicKeyMultikey(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L895](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L895): Method co.rsk.peg.Bridge.getPendingFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1420](https://github.com/rsksmart/rskj/blob/9c2ade78d6e30467176224008410dbcdc2f5b596/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1420): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 41 issues
* ![MAJOR][MAJOR] 16 major
* ![MINOR][MINOR] 16 minor
* ![INFO][INFO] 9 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [Bridge.java#L582](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L582): Method co.rsk.peg.Bridge.getBtcTransactionConfirmationsGetCost(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L592](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L592): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L594](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L594): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L596](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L596): Method co.rsk.peg.Bridge.getBtcTransactionConfirmations(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L678](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L678): Method co.rsk.peg.Bridge.getFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L751](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L751): Method co.rsk.peg.Bridge.getRetiringFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L814](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L814): Method co.rsk.peg.Bridge.addFederatorPublicKeyMultikey(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L815](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L815): Method co.rsk.peg.Bridge.addFederatorPublicKeyMultikey(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [Bridge.java#L895](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/Bridge.java#L895): Method co.rsk.peg.Bridge.getPendingFederatorPublicKeyOfType(Object[]) accesses list or array with constant index [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ACLI_CONSTANT_LIST_INDEX)
1. ![MAJOR][MAJOR] [BridgeSupport.java#L1420](https://github.com/rsksmart/rskj/blob/913781a607b451c0d9603ee314bd779a42f79579/rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java#L1420): Return an empty array instead of null. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by josedahlquist** [comment]
the comments will be tackled on a future PR

**Pull request closed without merging**